File: legacy/src/game/pinball-knight/gui/stack.ts
/**
 * THE SCREEN STACK — modality as game state instead of as DOM presence.
 *
 * `core.isSimPaused()` used to ask "is any of `shopEl / tavernEl /
 * cardReaderEl / menuEl` a live DOM node?". That worked, but it meant the
 * PAUSE CONTRACT — the single most important piece of state in the game, the
 * thing `simulate()` early-returns on — was stored in the document tree. A
 * screen that forgot to null its element left the game paused forever; one that
 * nulled it early unpaused the world underneath an open menu.
 *
 * Here a screen is a value on a stack. It pauses because it says it does.
 *
 * ## Why a stack and not a single `current`
 *
 * The tavern opens the vendor counter over itself; the menu can raise a confirm
 * over its own sheet; the haul screen hands off to the tavern. The DOM version
 * expressed this by layering `z-index`es and hoping the keyboard cascade in
 * `input/keymap.ts` matched — that cascade's comment already warns that the
 * ORDER of its checks is the design and reordering them is a behaviour change.
 * A stack makes that order explicit: input goes to the top, everything paints
 * bottom-up, and "most modal first" is a fact about the data rather than a
 * convention to be preserved by hand.
 */
import { state } from "../state";
import { setUiInputLive } from "./input";
import type { UiFrame } from "./im";

export interface UiScreen {
  /** Stable id — for debugging, `__gui.stack()`, and duplicate suppression. */
  id: string;
  /**
   * Freeze the simulation while this is open. HUDs and toasts are false; every
   * full-screen sheet is true.
   */
  pauses: boolean;
  /** Focus cursor, persisted across frames. Owned by the screen, not the frame. */
  focus: number;
  /** Scroll offset for the screen's main region, if it has one. */
  scroll: number;
  /**
   * The logical box this screen was authored for, in UI pixels.
   *
   * ── WHY EVERY SCREEN DOES NOT SHARE ONE SCALE ──
   * The UI layer is the pixel-pass grid, which on a desktop is 1600–1920 across.
   * Text on that grid is 8px Press Start 2P — one device pixel per font pixel,
   * which is as small as this UI can physically be, and it is what "the menus
   * are unreadable" means. The fix is to magnify by a whole number, and the
   * only question is BY HOW MUCH.
   *
   * A single global factor cannot answer that. The descent screen holds four
   * lines and would happily take 3x; the tavern is a 940x640 sheet and takes
   * none. Picking one number either leaves the sparse screens tiny or clips the
   * dense ones.
   *
   * So the screen declares the box it needs and the driver hands it the largest
   * INTEGER zoom at which that box still fits the grid. Sparse screens come out
   * big, dense screens come out unchanged, and neither can overflow. Screens
   * that omit this stay at 1x — which is the pre-existing behaviour, so adding
   * the field to a screen is opt-in and reversible.
   *
   * Whole numbers only, and that is not negotiable: the layer is nearest-
   * sampled pixel art and a x1.5 magnification makes every UI pixel
   * alternately 1 or 2 texels wide, which is the exact "game-wide mush" the
   * renderer's own integer-scale note in pixel-pass.ts was written to kill.
   */
  design?: {
    w: number;
    h: number;
    /**
     * Ceiling on the zoom this screen will accept, default `MAX_UI_ZOOM`.
     *
     * Without it a screen gets BIGGER on a bigger grid, in whole steps: the
     * sheets sit at 2x on a 1712-wide grid and would jump to 3x at 1920 — a 50%
     * growth from one browser-zoom notch. A cap is what turns "as big as it
     * fits" into "as big as it should be", and it costs nothing on the small
     * grids where the floor is binding anyway.
     */
    max?: number;
  };
  /** Paint one frame and handle its own input. */
  paint(f: UiFrame, self: UiScreen): void;
  /** Called when the screen is popped, for any teardown it owns. */
  onClose?(): void;
  /**
   * Handle `cancel` (Esc / B). Return true if handled; otherwise the screen is
   * popped. Screens with an armed confirm use this to disarm first.
   */
  onCancel?(self: UiScreen): boolean;
}

const stack: UiScreen[] = [];

/** Everything currently open, bottom-up. Painting order. */
export function screens(): readonly UiScreen[] {
  return stack;
}

/** The screen that owns input right now. */
export function top(): UiScreen | null {
  return stack.length ? stack[stack.length - 1] : null;
}

export function isOpen(id: string): boolean {
  return stack.some((s) => s.id === id);
}

/**
 * Recompute the pause flag `core.isSimPaused()` reads.
 *
 * Mirrored onto `state` rather than exported as a function so that core.ts can
 * consult it without a new import — the file is at its decomposition ratchet
 * (`core-boundary.test.ts`) and every other modal in that same expression is
 * already a `state` field, so this is the shape that file already expects. When
 * the DOM overlays are gone the whole expression collapses to this flag.
 */
/**
 * Whether the UI should own the keyboard right now.
 *
 * NOT "is any screen open". The HUD, the toasts and the floor map are all
 * screens and all stay open during play — gating capture on mere openness meant
 * the UI swallowed WASD, Space and every gameplay key the moment the HUD
 * existed, which is a game that renders perfectly and cannot be played.
 *
 * A screen wants the keyboard exactly when it pauses the world. That is the
 * same condition, and keeping it as one expression is what stops the two from
 * drifting apart.
 */
function syncPause(): void {
  state.uiPauses = stack.some((s) => s.pauses);
  // Arm the listeners THE INSTANT a pausing screen opens, not on the next
  // painted frame. Deferring it to the driver means every key pressed between
  // "the screen opened" and "the loop got round to rendering" is dropped — and
  // while a human cannot usually type that fast, a gamepad tap or anything that
  // opens a screen programmatically absolutely can. It presented as a flaky
  // off-by-one in the focus cursor.
  setUiInputLive(state.uiPauses);
}

export function push(s: UiScreen): void {
  // Re-opening what is already on top is a no-op, not a second copy. The DOM
  // version guarded this with `if (state.menuEl) return` in every opener; here
  // it is one rule in one place.
  if (top()?.id === s.id) return;
  stack.push(s);
  syncPause();
}

export function pop(): UiScreen | null {
  const s = stack.pop() ?? null;
  s?.onClose?.();
  syncPause();
  return s;
}

/**
 * Remove exactly ONE screen, wherever it sits. Safe if it was never there.
 *
 * The counterpart to `close()`, and the distinction is load-bearing. `close()`
 * truncates — it drops the named screen AND everything above it — which is
 * right for a modal sheet, because closing a parent must close the child it
 * raised. It is WRONG for the always-on layers (the HUD, the toasts, the touch
 * pad, the tavern's station prompt), which sit at the bottom and are
 * independent of whatever is stacked over them.
 *
 * Measured 2026-07-29: the tavern raised its station prompt and then called
 * `hideDungeonHud(true)` → `close("hud")`, which found the HUD at index 0 and
 * truncated the whole stack — silently taking the prompt and the lobby board
 * with it. The tavern rendered perfectly and had no interface at all.
 */
export function remove(id: string): void {
  const i = stack.findIndex((s) => s.id === id);
  if (i < 0) return;
  stack[i].onClose?.();
  stack.splice(i, 1);
  syncPause();
}

/** Pop until `id` is gone — it AND everything above it. See `remove()`. */
export function close(id: string): void {
  const i = stack.findIndex((s) => s.id === id);
  if (i < 0) return;
  for (let n = stack.length - 1; n >= i; n--) stack[n].onClose?.();
  stack.length = i;
  syncPause();
}

/** Drop everything — run teardown, level change, dispose. */
export function clearScreens(): void {
  for (let n = stack.length - 1; n >= 0; n--) stack[n].onClose?.();
  stack.length = 0;
  syncPause();
}

File: crates/pk-gui/src/stack.rs
//! The screen stack — `legacy/src/game/pinball-knight/gui/stack.ts`.
//!
//! Modality as data: input goes to the top, everything paints bottom-up, and a
//! screen pauses the world because it SAYS it does (`pauses`), not because a
//! DOM node happens to exist. The `remove` / `close` distinction is
//! load-bearing: `close` truncates (a parent closing must close the child it
//! raised); `remove` takes exactly one screen wherever it sits — the always-on
//! layers (station prompt, HUD) live at the bottom and must never take the
//! stack with them (stack.ts's 2026-07-29 war story).

/// The logical box a screen was authored for. The driver hands the screen the
/// largest INTEGER zoom at which the box still fits the grid, capped by `max`.
#[derive(Clone, Copy, Debug)]
pub struct Design {
    pub w: f64,
    pub h: f64,
    pub max: u32,
}

#[derive(Clone, Debug)]
pub struct ScreenEntry<Id: Copy + Eq> {
    pub id: Id,
    /// Freeze the simulation while this is open. HUDs and prompts are false;
    /// every full-screen sheet is true.
    pub pauses: bool,
    /// Focus cursor, persisted across frames. Owned by the screen.
    pub focus: i64,
    /// Scroll offset for the screen's main region, if it has one.
    pub scroll: f64,
    /// This screen paints something that moves, so it needs frames even when
    /// no input arrives.
    ///
    /// The shell skips `paint_stack` entirely on a quiet frame over an
    /// unchanged stack — that skip is most of the game's idle cost and is not
    /// negotiable (it was worth a measured 36→14 fps when it regressed). A
    /// screen that animates is the one case where "nothing arrived" does not
    /// mean "nothing changed", so it opts out here rather than by weakening
    /// the skip for everyone.
    pub animates: bool,
    pub design: Option<Design>,
}

impl<Id: Copy + Eq> ScreenEntry<Id> {
    pub fn new(id: Id, pauses: bool) -> Self {
        Self {
            id,
            pauses,
            focus: 0,
            scroll: 0.0,
            animates: false,
            design: None,
        }
    }

    pub fn with_design(mut self, w: f64, h: f64, max: u32) -> Self {
        self.design = Some(Design { w, h, max });
        self
    }

    /// Mark this screen as animating — see [`ScreenEntry::animates`].
    pub fn animating(mut self) -> Self {
        self.animates = true;
        self
    }
}

#[derive(Default)]
pub struct UiStack<Id: Copy + Eq> {
    entries: Vec<ScreenEntry<Id>>,
}

impl<Id: Copy + Eq> UiStack<Id> {
    pub fn new() -> Self {
        Self {
            entries: Vec::new(),
        }
    }

    /// Everything currently open, bottom-up. Painting order.
    pub fn screens(&self) -> &[ScreenEntry<Id>] {
        &self.entries
    }

    pub fn screens_mut(&mut self) -> &mut [ScreenEntry<Id>] {
        &mut self.entries
    }

    pub fn len(&self) -> usize {
        self.entries.len()
    }

    pub fn is_empty(&self) -> bool {
        self.entries.is_empty()
    }

    /// The screen that owns input right now.
    pub fn top(&self) -> Option<&ScreenEntry<Id>> {
        self.entries.last()
    }

    pub fn is_open(&self, id: Id) -> bool {
        self.entries.iter().any(|s| s.id == id)
    }

    /// Whether the UI should own the keyboard right now — NOT "is any screen
    /// open": the prompt/HUD are open all run and must not eat WASD.
    pub fn pauses(&self) -> bool {
        self.entries.iter().any(|s| s.pauses)
    }

    /// Does any open screen need frames without input?
    ///
    /// Any, not just the top: an animation under a modal keeps running, which
    /// is what the oracle's single RAF does. The shell consults this to decide
    /// whether a quiet frame may be skipped.
    pub fn animates(&self) -> bool {
        self.entries.iter().any(|s| s.animates)
    }

    /// Re-opening what is already on top is a no-op, not a second copy.
    pub fn push(&mut self, s: ScreenEntry<Id>) {
        if self.top().map(|t| t.id == s.id).unwrap_or(false) {
            return;
        }
        self.entries.push(s);
    }

    pub fn pop(&mut self) -> Option<ScreenEntry<Id>> {
        self.entries.pop()
    }

    /// Remove exactly ONE screen, wherever it sits. Safe if it was never there.
    pub fn remove(&mut self, id: Id) {
        if let Some(i) = self.entries.iter().position(|s| s.id == id) {
            self.entries.remove(i);
        }
    }

    /// Pop until `id` is gone — it AND everything above it.
    pub fn close(&mut self, id: Id) {
        if let Some(i) = self.entries.iter().position(|s| s.id == id) {
            self.entries.truncate(i);
        }
    }

    pub fn clear(&mut self) {
        self.entries.clear();
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    /// `animates()` asks about ANY open screen, not the top one.
    ///
    /// Same shape as `pauses()`, and for the same reason: a spinning wheel
    /// under a confirm box still has to be painted, so the question the shell
    /// asks is "does anything on this stack need frames", not "does the screen
    /// receiving input need them".
    #[test]
    fn a_screen_that_animates_makes_the_whole_stack_animate() {
        let mut s: UiStack<u8> = UiStack::new();
        assert!(!s.animates(), "an empty stack asks for nothing");
        s.push(ScreenEntry::new(0, false));
        assert!(!s.animates(), "a static screen asks for nothing");
        s.push(ScreenEntry::new(1, true).animating());
        assert!(s.animates());
        // Buried under a modal, it still wants frames.
        s.push(ScreenEntry::new(2, true));
        assert!(s.animates());
        s.remove(1);
        assert!(!s.animates(), "and stops wanting them once it closes");
    }

    /// Screens are static unless they say otherwise.
    ///
    /// The flag has to default false or every existing screen would silently
    /// defeat the shell's quiet-frame skip, which is most of the idle cost.
    #[test]
    fn a_screen_is_static_until_it_asks_not_to_be() {
        assert!(!ScreenEntry::new(0u8, true).animates);
        assert!(ScreenEntry::new(0u8, true).animating().animates);
        // …and it composes with the design box, in either order.
        assert!(
            ScreenEntry::new(0u8, true)
                .with_design(600.0, 420.0, 2)
                .animating()
                .animates
        );
    }

    #[test]
    fn remove_takes_one_screen_and_close_truncates() {
        // The stack.ts war story: remove("hud") must not take the prompt with it.
        let mut s: UiStack<u8> = UiStack::new();
        s.push(ScreenEntry::new(0, false)); // hud
        s.push(ScreenEntry::new(1, false)); // prompt
        s.push(ScreenEntry::new(2, true)); // sheet
        s.remove(0);
        assert_eq!(s.len(), 2);
        assert!(s.is_open(1) && s.is_open(2));

        s.push(ScreenEntry::new(3, true));
        s.close(2); // closing the sheet takes the confirm it raised
        assert_eq!(s.len(), 1);
        assert!(s.is_open(1));
    }

    #[test]
    fn pauses_follows_the_flag_not_openness() {
        let mut s: UiStack<u8> = UiStack::new();
        s.push(ScreenEntry::new(0, false));
        assert!(!s.pauses());
        s.push(ScreenEntry::new(1, true));
        assert!(s.pauses());
    }

    #[test]
    fn push_dedupes_only_against_the_top() {
        let mut s: UiStack<u8> = UiStack::new();
        s.push(ScreenEntry::new(0, false));
        s.push(ScreenEntry::new(0, false));
        assert_eq!(s.len(), 1);
    }
}
